File: 范例/Accpt/16xx/p1649.pas
PROGRAM p1649;

CONST
	maxn=200;
	change			:array[1..4,1..2]of Integer=
	((1,0),(0,1),(-1,0),(0,-1));

TYPE
	node=record
	a,b		:Integer;
	end;

VAR
map			:array[1..maxn,1..maxn]of Char;
minway			:array[1..maxn,1..maxn]of Integer;
w,h			:Integer;

PROCEDURE Readin;
var
i,j			:Integer;
begin
readln(h,w);
for i:=1 to h do
begin
for j:=1 to w do
read(map[i,j]);
readln;
end;
end;

PROCEDURE Search(x,y:Integer);
var
s,t,i,p,q 		:Integer;
line			:array[1..40000]of Node;
begin
with line[1] do
begin
a:=x;
b:=y;
end;
s:=1;
t:=1;
repeat
for i:=1 to 4 do
begin
p:=line[s].a+change[i,1];
q:=line[s].b+change[i,2];
if (p<1)or(p>h)or(q<1)or(q>w)or(map[p,q]='#') then continue;
if (map[p,q]='x') then
begin
if (minway[line[s].a,line[s].b]+2<minway[p,q]) then
begin
inc(t);
line[t].a:=p;
line[t].b:=q;
minway[p,q]:=minway[line[s].a,line[s].b]+2;
end;
end else
if minway[line[s].a,line[s].b]+1<minway[p,q] then
begin
inc(t);
line[t].a:=p;
line[t].b:=q;
minway[p,q]:=minway[line[s].a,line[s].b]+1;
end;
end;
inc(s);
until s>t;
end;

PROCEDURE main;
var
i,j,a,b,ansa,ansb	:Integer;
begin
for i:=1 to h do
for j:=1 to w do
minway[i,j]:=maxint;

for i:=1 to h do
for j:=1 to w do
if map[i,j]='r' then
begin
minway[i,j]:=0;
Search(i,j);
end else
if map[i,j]='a' then
begin
ansa:=i;
ansb:=j;
end;
if minway[ansa,ansb]<maxint then
	writeln(minway[ansa,ansb])
else writeln('Poor ANGEL has to stay in the prison all his life.');
end;

BEGIN
//assign(input,'p.in');
//reset(input);
while not eof do
begin
readin;
main;
end;
END.

File: Vol.07(1600-1699)/#1649 Rescue/p1649.cpp
// 2882769 2008-05-01 17:21:00 Accepted 1649 C++ 00:00.30 476K 呆滞的慢板 
// BFS，就是 WA... 中间有逻辑错误，然后换了种不用队列的方法...
// 慢死了，不是一个量级的复杂度... 

#include <cstdio>
#include <cstring>
#include <cstdlib>

int D[4][2] = { { 0,  1 }, {  1, 0 },
                { 0, -1 }, { -1, 0 } };

int     X,  Y,  step;
bool    change, keep, finish;
char    M[200][210];
char    N[200][210];

int main() {
    
    while( scanf( "%d%d ", &X, &Y ) != EOF ) {
        
        memset( M, 0, sizeof( M ) );
        
        for( int i = 1; i <= X; ++i )
            gets( M[i] + 1 );
        
        for( step = finish = 0; true; ++step ) {
            
            change = false;
            memcpy( N, M, sizeof( M ) );
            
            for( int i = 1; i <= X; ++i ) {
                
                for( int j = 1; j <= Y; ++j ) {
                    
                    if( N[i][j] == 'r' ) {
                        change = true;
                        keep = false;
                        for( int d = 0; d < 4; ++d ) {
                            switch( N[i+D[d][0]][j+D[d][1]] ) {
                                
                                case 'x':
                                    M[i+D[d][0]][j+D[d][1]] = '.';
                                    keep = true;
                                break;
                                
                                case '.':
                                    M[i+D[d][0]][j+D[d][1]] = 'r';
                                break;
                                
                                case 'a':
                                    finish = true;
                                break;
                                
                            }
                        }
                        
                        if( !keep )   M[i][j] = '#';
                        
                        if( finish )  break;
                        
                    }
                    
                }
                
                if( finish )  break;
                
            }
            
            if( !change || finish ) break;
        }
            
        if( finish )printf( "%d\n", step + 1 );
        else        puts( "Poor ANGEL has to stay in the prison all his life." );

    }
    
}
